client/core/trade.go

```diff
@@ -1405,60 +1405,95 @@ func (c *Core) tick(t *trackedTrade) (assetMap, error) {
-	errs := newErrorSet(t.dc.acct.host + " tick: ")
+	assets := make(assetMap) // callers expect non-nil map even on error
 
 	// Check all matches for and resend pending requests as necessary.
 	c.resendPendingRequests(t)
 
 	// Make sure we have a redemption fee suggestion cached.
 	t.cacheRedemptionFeeSuggestion()
 
-	// Check all matches and send swap, redeem or refund as necessary.
 	var sent, quoteSent, received, quoteReceived uint64
-	for _, match := range t.matches {
+	checkMatch := func(match *matchTracker) error {
 		side := match.Side
 		if (side == order.Maker && match.Status >= order.MakerRedeemed) ||
 			(side == order.Taker && match.Status >= order.MatchComplete) {
-			continue
+			return nil
 		}
 		if match.Address == "" {
-			continue // a cancel order match
+			return nil // a cancel order match
 		}
 		if !t.matchIsActive(match) {
-			continue // either refunded or revoked requiring no action on this side of the match
+			return nil // either refunded or revoked requiring no action on this side of the match
 		}
-		switch {
-		case t.isSwappable(c.ctx, match):
+
+		// The trackedTrade mutex is locked, so we must not hang forever. Give
+		// this a generous timeout because it may be necessary to retrieve full
+		// blocks, and catch timeout/shutdown after each check.
+		ctx, cancel := context.WithTimeout(c.ctx, 30*time.Second)
+		defer cancel()
+
+		if t.isSwappable(ctx, match) {
 			c.log.Debugf("Swappable match %s for order %v (%v)", match, t.ID(), side)
 			swaps = append(swaps, match)
 			sent += match.Quantity
 			quoteSent += calc.BaseToQuote(match.Rate, match.Quantity)
+			return nil
+		}
+		if ctx.Err() != nil { // may be here because of timeout or shutdown
+			return ctx.Err()
+		}
 
-		case t.isRedeemable(c.ctx, match):
+		if t.isRedeemable(ctx, match) {
 			c.log.Debugf("Redeemable match %s for order %v (%v)", match, t.ID(), side)
 			redeems = append(redeems, match)
 			received += match.Quantity
 			quoteReceived += calc.BaseToQuote(match.Rate, match.Quantity)
+			return nil
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 
 		// Check refundability before checking if to start finding redemption.
 		// Ensures that redemption search is not started if locktime has expired.
 		// If we've already started redemption search for this match, the search
 		// will be aborted if/when auto-refund succeeds.
-		case t.isRefundable(match):
+		if t.isRefundable(ctx, match) {
 			c.log.Debugf("Refundable match %s for order %v (%v)", match, t.ID(), side)
 			refunds = append(refunds, match)
+			return nil
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 
-		case t.shouldBeginFindRedemption(c.ctx, match):
+		if t.shouldBeginFindRedemption(ctx, match) {
 			c.log.Debugf("Ready to find counter-party redemption for match %s, order %v (%v)", match, t.ID(), side)
-			t.findMakersRedemption(match)
+			t.findMakersRedemption(c.ctx, match) // async search, not on the timeout
+			return nil
+		}
+
+		return ctx.Err()
```

Comment:
This comes down to what were arguing about in https://github.com/decred/dcrdex/pull/1732#discussion_r935502352

The whole reason this looks and feels weird is because these `is` functions aren't written to return and error.  Normally we'd have a pattern like this where the error from the function may be one of `context.DeadlineExceeded` or `context.Canceled`:

```go
err := doSomething(ctx)
if err != nil {
	if errors.Is(err, context.DeadlineExceeded) {
		 fmt.Println("timeout")
	} else if errors.Is(err, context.Canceled) {
		fmt.Println("shutdown")
	}
	return err
}
```

I didn't want to refactor those `is` helpers so much, so I'm just checking if the reason for the `false` return is because of either context timeout/shutdown **rather than** a legitimate `false` from successful checks.

In the previous revision with the `switch`-`case` it was very possible that one of the cases was `false` for one of these bad reasons (real error) but the `switch` let it go to check the next `case` because that's how a `switch` works.  These changes are just a minor improvement to prevent that erroneous flow and instead bail out to try again.

Reply:
It looks like that would be a change in behavior and that we couldn't return any of the errors that we're currently logging in these methods. The reason those errors are only logged is because an error on one match can't stop others from proceeding, which it appears would no longer be the case in this new scheme. That's why the single boolean return makes sense here. There was no error that should have change the behavior of the caller, so there was no need to return one.

Reply:
> The reason those errors are only logged is because an error on one match can't stop others from proceeding, which it appears would no longer be the case in this new scheme.

Well that's another reason I didn't make that change.  But I didn't say every internal error would need to return an error, just that the pattern of checking the context in the caller instead of an error exists because of this current design.  Nothing needs to change IMO.